crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,262 @@
+    pub fn unchecked_set_count(&mut self, count: u32) -> u32 {
+        self.count = count;
+        self.count
+    }
+
+    /// Splits this `ItemStack` in half, returning the
+    /// removed half. If the amount is odd, `self`
+    /// will be left with the least items. Returns the taken
+    /// half.
+    pub fn take_half(&mut self) -> ItemStack {
+        self.take((self.count as f64 / 2 as f64).ceil() as u32)
+            .unwrap()
+    }
+
+    /// Splits this `ItemStack` by removing the
+    /// specified amount. Returns the taken part.
+    pub fn take(&mut self, amount: u32) -> Result<ItemStack, ItemStackError> {
+        if self.count <= amount {
+            return Err(if self.count == amount {
+                ItemStackError::EmptyStack
+            } else {
+                ItemStackError::NotEnoughAmount
+            });
+        }
+        let count_left: u32 = self.count - amount;
+        let taken = ItemStack {
+            count: amount,
+            ..self.clone()
+        };
+        self.count = count_left as u32;
+        Ok(taken)
+    }
+
+    /// Merges another `ItemStack` with this one.
+    pub fn merge_with(&mut self, other: &mut Self) -> bool {
+        if !self.has_same_type_and_damage(other) {
+            return false;
+        }
+        let new_count = (self.count + other.count).min(self.item.stack_size());
+        let amount_added = new_count - self.count;
+        self.count = new_count;
+        other.count -= amount_added;
+        true
+    }
+
+    /// Transfers up to `n` items to `other`.
+    pub fn transfer_to(&mut self, n: u32, other: &mut Self) -> Result<(), ItemStackError> {
+        if self.count <= n {
+            return Err(if self.count == n {
+                ItemStackError::EmptyStack
+            } else {
+                ItemStackError::NotEnoughAmount
+            });
+        }
+        let max_transfer = other.item.stack_size().saturating_sub(other.count);
+        let transfer = max_transfer.min(self.count).min(n);
+        if other.count + transfer > i32::MAX as u32 {
+            return Err(ItemStackError::ClientOverflow);
+        }
+        self.count -= transfer;
+        other.count += transfer;
+        Ok(())
+    }
+
+    /// Damages the item by the specified amount.
+    /// If this function returns `true`, then the item is broken.
+    pub fn damage(&mut self, amount: u32) -> bool {
+        if self.meta.is_none() {
+            return false;
+        }
+        match &mut self.meta.clone().unwrap().damage {
+            Some(damage) => {
+                *damage += amount;
+                if let Some(durability) = self.item.durability() {
+                    *damage >= durability
+                } else {
+                    false
+                }
+            }
+            None => false,
+        }
+    }
+}
+
+/// An error type that may be returned when performing
+/// operations over an `ItemStack`.
+#[derive(Debug)]
+pub enum ItemStackError {
```

Comment:
We should implement `Display` and `Error` for this type to conform to Rust API guidelines.